ml4h/models.py

```diff
@@ -842,13 +868,409 @@ def __call__(self, x: Tensor, intermediates: Dict[TensorMap, List[Tensor]], _) -
+            pool_z: int,
+            **kwargs,
+    ):
+        self.tensor_map = tensor_map
+        if not self.can_apply():
+            return
+        dimension = self.tensor_map.axes()
+
+        # list of filter dimensions should match the total number of convolutional layers
+        x_filters = _repeat_dimension(conv_width if dimension == 2 else conv_x, len(conv_layers)+len(dense_blocks))
+        y_filters = _repeat_dimension(conv_y, len(conv_layers)+len(dense_blocks))
+        z_filters = _repeat_dimension(conv_z, len(conv_layers)+len(dense_blocks))
+
+        #self.preprocess_block = PreprocessBlock(['rotate'], [0.3])
+        self.res_block = ResidualBlock(
+            dimension=dimension, filters_per_conv=conv_layers, conv_layer_type=conv_type, conv_x=x_filters[:len(conv_layers)],
+            conv_y=y_filters[:len(conv_layers)], conv_z=z_filters[:len(conv_layers)], activation=activation, normalization=conv_normalize,
+            regularization=conv_regularize, regularization_rate=conv_regularize_rate, dilate=conv_dilate,
+        )
+
+        self.dense_blocks = [
+            DenseConvolutionalBlock(
+                dimension=dimension, conv_layer_type=conv_type, filters=filters, conv_x=[x] * block_size, conv_y=[y] * block_size,
+                conv_z=[z]*block_size, block_size=block_size, activation=activation, normalization=conv_normalize,
+                regularization=conv_regularize, regularization_rate=conv_regularize_rate,
+            ) for filters, x, y, z in zip(dense_blocks, x_filters[len(conv_layers):], y_filters[len(conv_layers):], z_filters[len(conv_layers):])
+        ]
+        self.pools = _pool_layers_from_kind_and_dimension(dimension, pool_type, len(dense_blocks) + 1, pool_x, pool_y, pool_z)
+        self.fully_connected = FullyConnectedBlock(
+            widths=dense_layers,
+            activation=activation,
+            normalization=dense_normalize,
+            regularization=dense_regularize,
+            regularization_rate=dense_regularize_rate,
+            name=f'embed_{self.tensor_map.name}',
+        ) if dense_layers else None
+
+    def can_apply(self):
+        return self.tensor_map.axes() > 1
+
+    def __call__(self, x: Tensor, intermediates: Dict[TensorMap, List[Tensor]]) -> Tensor:
+        if not self.can_apply():
+            return x
+        #x = self.preprocess_block(x)  # TODO: upgrade to tensorflow 2.3+
+        x = self.res_block(x)
+        intermediates[self.tensor_map].append(x)
+        for i, (dense_block, pool) in enumerate(zip(self.dense_blocks, self.pools)):
+            x = pool(x)
             x = dense_block(x)
+            intermediates[self.tensor_map].append(x)
+        if self.fully_connected:
+            x = Flatten()(x)
+            x = self.fully_connected(x)
+            intermediates[self.tensor_map].append(x)
+        return x
+
+
+class ConvDecoderBlock:
+    def __init__(
+            self,
+            *,
+            tensor_map: TensorMap,
```

Comment:
If I understand right, this is the output tensormap for a model, so maybe call it `tensor_map_out`

Reply:
we could, but it's kinda nice to have just tensor_map for Blocks that can be encoders or decoders not to have to specify `_in` or `_out`